### PR TITLE
[Windows] Prepare to add/remove views

### PR DESCRIPTION
**Pull request by loic-sharma**
_This is a refactoring with no semantic changes._

Prepare the Windows embedder to add/remove views. In a multi-view world, the platform thread can remove a view while the raster thread presenting to the same view. A lock will be introduced to ensure this does not happen. This shuffles the code around so that the engine can acquire the lock before the compositor presents to the view.

Prepares for https://github.com/flutter/flutter/issues/138179
Part of https://github.com/flutter/flutter/issues/142845

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/compositor_opengl_unittests.cc [hunk 1]
```diff
@@ -185,53 +182,7 @@ TEST_F(CompositorOpenGLTest, PresentEmpty) {
   EXPECT_CALL(*surface(), IsValid).WillRepeatedly(Return(true));
   EXPECT_CALL(*surface(), MakeCurrent).WillOnce(Return(true));
   EXPECT_CALL(*surface(), SwapBuffers).WillOnce(Return(true));
-  EXPECT_TRUE(compositor.Present(view()->view_id(), nullptr, 0));
-}
-
-TEST_F(CompositorOpenGLTest, HeadlessPresentIgnored) {
```
Comment (loic-sharma) [inline shell/platform/windows/compositor_opengl_unittests.cc:191]: The following tests were all replaced by `WindowsTest.PresentHeadless`:

1. `CompositorOpenGLTest.HeadlessPresentIgnored`
2. `CompositorOpenGLTest.UnknownViewIgnored`
3. `CompositorSoftwareTest.HeadlessPresentIgnored`
4. `CompositorSoftwareTest.UnknownViewIgnored`

(Presenting to the implicit view while in headless mode is effectively the same thing as presenting to an unknown view)

File: shell/platform/windows/fixtures/main.dart [hunk 2]
```diff
@@ -315,9 +315,49 @@ void drawHelloWorld() {
       ..addPicture(ui.Offset.zero, picture)
       ..pop();
 
-    ui.window.render(sceneBuilder.build());
+    ui.PlatformDispatcher.instance.implicitView?.render(sceneBuilder.build());
```
Comment (loic-sharma) [inline shell/platform/windows/fixtures/main.dart:318]: The `window` global is deprecated and replaced by `PlatformDispatcher.instance.implicitView`. This isn't necessary for this PR, but removes a usage of a deprecated member :)

File: shell/platform/windows/flutter_windows_unittests.cc [hunk 3]
```diff
@@ -481,41 +546,28 @@ TEST_F(WindowsTest, Lifecycle) {
...
-          ASSERT_FALSE(Dart_IsError(handle));
-          auto value = tonic::DartConverter<std::string>::FromDart(handle);
-          EXPECT_EQ(value, "Success");
-          done = true;
-          latch.Signal();
-        }));
-
-    ViewControllerPtr controller{builder.Run()};
-    ASSERT_NE(controller, nullptr);
+  auto& context = GetContext();
```
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_unittests.cc:549]: The extra thread was unnecessary for this test. Consider reviewing this with whitespace changes disabled: https://github.com/flutter/engine/pull/51908/files?w=1#diff-f0b8369989e179cde8da71010c79f2c542f16c55f91e9d9aeae1a3fe769cf83dR548

File: shell/platform/windows/compositor_opengl.cc [hunk 4]
```diff
@@ -92,14 +92,9 @@ bool CompositorOpenGL::CollectBackingStore(const FlutterBackingStore* store) {
   return true;
 }
 
-bool CompositorOpenGL::Present(FlutterViewId view_id,
+bool CompositorOpenGL::Present(FlutterWindowsView* view,
                                const FlutterLayer** layers,
                                size_t layers_count) {
```
Comment (cbracken) [inline shell/platform/windows/compositor_opengl.cc:97]: Depending on how paranoid you feel, consider tossing in an `FML_DCHECK(view != nullptr)` in the updated methods to shout should anyone ever violate preconditions.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

**Review by yaakovschectman — APPROVED** [review]

File: shell/platform/windows/compositor_opengl.cc (hunk 4 above)
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc:97]: Added!

**Pull request merged**